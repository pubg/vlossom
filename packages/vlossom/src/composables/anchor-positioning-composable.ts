import { Ref, onBeforeMount, ref, nextTick } from 'vue';
import { utils } from '@/utils';
import type { AttachInfo, Placement, Align } from '@/declaration';

export function usePositioning(anchor: Ref<HTMLElement>, attachment: Ref<HTMLElement>) {
    const isVisible = ref(false);
    const computedPlacement: Ref<Exclude<Placement, 'middle'> | null> = ref(null);
    let throttledComputePosition: ((...args: any) => any) | null = null;
    let resizeObserver: ResizeObserver | null = null;

    function getX(align: Align, left: number, right: number, width: number, attachmentWidth: number) {
        switch (align) {
            case 'start':
                return left;
            case 'end':
                return right - attachmentWidth;
            case 'center':
                return left + width / 2 - attachmentWidth / 2;
            default:
                throw Error('Align is Invalid Value');
        }
    }

    function getY(align: Align, top: number, bottom: number, height: number, attachmentHeight: number) {
        switch (align) {
            case 'start':
                return top;
            case 'end':
                return bottom - attachmentHeight;
            case 'center':
                return top + height / 2 - attachmentHeight / 2;
            default:
                throw Error('Align is Invalid Value');
        }
    }

    function computePosition({ placement = 'top', align = 'center', margin = 0, followWidth = false }: AttachInfo) {
        if (!anchor.value || !attachment.value) {
            return;
        }

        const { top, right, bottom, left, width, height } = utils.dom.getClientRect(anchor.value);
        const { width: attachmentWidth, height: attachmentHeight } = utils.dom.getClientRect(attachment.value);

        // Change placements when there are no spaces in the viewport.
        if (placement === 'bottom' && bottom + attachmentHeight > window.innerHeight) {
            computedPlacement.value = window.innerHeight - bottom < top ? 'top' : 'bottom';
        } else if (placement === 'top' && top - attachmentHeight < 0) {
            computedPlacement.value = window.innerHeight - bottom > top ? 'bottom' : 'top';
        } else if (placement === 'left' && left - attachmentWidth < 0) {
            computedPlacement.value = window.innerWidth - right > left ? 'right' : 'left';
        } else if (placement === 'right' && right + attachmentWidth > window.innerWidth) {
            computedPlacement.value = window.innerWidth - right < left ? 'left' : 'right';
        } else {
            computedPlacement.value = placement;
        }

        let x: number;
        let y: number;

        switch (computedPlacement.value) {
            case 'top':
                x = getX(align, left, right, width, attachmentWidth);
                y = top - attachmentHeight - margin;
                break;
            case 'right':
                x = right + margin;
                y = getY(align, top, bottom, height, attachmentHeight);
                break;
            case 'bottom':
                x = getX(align, left, right, width, attachmentWidth);
                y = bottom + margin;
                break;
            case 'left':
                x = left - attachmentWidth - margin;
                y = getY(align, top, bottom, height, attachmentHeight);
                break;
            default:
                throw Error('Placement is Invalid Value');
        }

        const scrollLeft = window.scrollX || document.documentElement.scrollLeft || 0;
        const scrollTop = window.scrollY || document.documentElement.scrollTop || 0;

        attachment.value.style.left = `${x + scrollLeft}px`;
        attachment.value.style.top = `${y + scrollTop}px`;

        if (followWidth) {
            attachment.value.style.width = `${width}px`;
        }
    }

    function appear(attachInfo: AttachInfo = {}) {
        isVisible.value = true;

        nextTick(() => {
            try {
                attachment.value.style.display = 'block';
                attachment.value.style.position = 'absolute';
                computePosition(attachInfo);

                throttledComputePosition = utils.function.throttle(computePosition.bind(null, attachInfo), 30);

                const hasResizeObserver = window && window.ResizeObserver !== undefined;
                if (hasResizeObserver) {
                    resizeObserver = new ResizeObserver(throttledComputePosition);
                    resizeObserver.observe(anchor.value);
                    resizeObserver.observe(attachment.value);
                }

                document.addEventListener('scroll', throttledComputePosition, true);
                window.addEventListener('resize', throttledComputePosition, true);
            } catch (error: any) {
                utils.log.error('anchor positioning', error);
            }
        });
    }

    function disappear() {
        if (throttledComputePosition) {
            resizeObserver?.disconnect();
            document.removeEventListener('scroll', throttledComputePosition, true);
            window.removeEventListener('resize', throttledComputePosition, true);
        }
        isVisible.value = false;
    }

    return {
        isVisible,
        computedPlacement,
        appear,
        disappear,
    };
}

export function useOverlay() {
    onBeforeMount(() => {
        if (document.getElementById('vs-overlay')) {
            return;
        }
        const overlay = document.createElement('div');
        overlay.setAttribute('id', 'vs-overlay');

        overlay.style.position = 'absolute';
        overlay.style.top = '0';
        overlay.style.left = '0';
        overlay.style.zIndex = '10000';

        document.body.appendChild(overlay);
    });
}
