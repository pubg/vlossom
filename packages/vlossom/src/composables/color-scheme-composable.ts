import type { Ref } from 'vue';
import type { ColorScheme, VsComponent } from '@/declaration/types';

import { computed } from 'vue';
import { store } from '@/store';

export function useColorScheme(component: VsComponent, colorScheme: Ref<ColorScheme | undefined>) {
    const computedColorScheme = computed(
        () =>
            colorScheme.value ||
            store.getStore().globalColorScheme[component] ||
            store.getStore().globalColorScheme.default ||
            'default',
    );

    return {
        computedColorScheme,
    };
}
