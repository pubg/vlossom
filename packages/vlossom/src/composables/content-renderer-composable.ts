import { Component, h, toRaw } from 'vue';
import { VsContentRenderer } from '@/nodes';

export function useContentRenderer() {
    function getRenderedContent(content: string | Component) {
        if (typeof content === 'string') {
            return h(VsContentRenderer, { content });
        }

        // vue component
        return toRaw(content);
    }

    return { getRenderedContent };
}
