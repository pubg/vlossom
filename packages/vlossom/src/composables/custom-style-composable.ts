import { computed, ComputedRef, Ref } from 'vue';
import { utils } from '@/utils';
import { store } from '@/store';

import type { VsComponent } from '@/declaration';

export function useCustomStyle<T extends { [key: string]: any }>(component: VsComponent, styleSet: Ref<string | T>) {
    const styles: ComputedRef<T> = computed(() => {
        if (!styleSet.value) {
            return {} as T;
        }

        if (typeof styleSet.value === 'string') {
            return (store.getStyleSet(component, styleSet.value) || {}) as T;
        }

        return styleSet.value;
    });

    const customProperties = computed(() =>
        Object.entries(styles.value).reduce((acc, [key, value]) => {
            acc[`--${utils.string.pascalToKebab(component)}-${key}`] = value;
            return acc;
        }, {} as { [key: string]: any }),
    );

    return {
        customProperties,
    };
}
