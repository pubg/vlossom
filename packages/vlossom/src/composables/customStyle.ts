import type { Ref } from 'vue';
import type { StyleSet } from '@/declaration/types';

import { computed, ComputedRef, ref } from 'vue';
import { kebabToPascal } from '@/utils';

const registeredStyleSet: Ref<StyleSet> = ref({});

export function registerStyleSet(newStyleSet: StyleSet) {
    Object.entries(newStyleSet).forEach(([key, value]) => {
        const styleSet = registeredStyleSet.value[key as keyof StyleSet];

        if (!styleSet) {
            registeredStyleSet.value[key as keyof StyleSet] = value;
        } else {
            Object.assign(styleSet, value);
        }
    });
}

export function useCustomStyle<T extends { [key: string]: any }>(styleSet: Ref<string | T>, prefix: string) {
    const styles: ComputedRef<T> = computed(() => {
        if (!styleSet.value) {
            return {} as T;
        }

        if (typeof styleSet.value === 'string') {
            const preDefinedStyleSet =
                registeredStyleSet.value[('vs' + kebabToPascal(prefix)) as keyof StyleSet]?.[styleSet.value];

            return (preDefinedStyleSet ?? {}) as T;
        }

        return styleSet.value;
    });

    const customProperties = computed(() =>
        Object.entries(styles.value).reduce((acc, [key, value]) => {
            acc[`--vs-${prefix}-${key}`] = value;
            return acc;
        }, {} as { [key: string]: any }),
    );

    return {
        customProperties,
    };
}
