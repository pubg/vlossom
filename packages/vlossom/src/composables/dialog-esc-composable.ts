import { store } from '@/stores';
import { Ref, onBeforeUnmount, onMounted } from 'vue';

export function useDialogEsc(id: string, closeOnEsc: Ref<boolean>, onEsc: () => void) {
    function onPressEsc(event: KeyboardEvent) {
        if (closeOnEsc.value === false) {
            return;
        }

        if (event.key === 'Escape' && store.dialog.getTopId() === id) {
            onEsc();
        }
    }

    onMounted(() => {
        document.addEventListener('keydown', onPressEsc);
        if (closeOnEsc.value) {
            store.dialog.push(id);
        }
    });

    onBeforeUnmount(() => {
        document.removeEventListener('keydown', onPressEsc);
        if (closeOnEsc.value) {
            store.dialog.pop();
        }
    });
}
