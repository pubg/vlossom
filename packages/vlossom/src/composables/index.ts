export * from './anchor-positioning-composable';
export * from './color-scheme-composable';
export * from './form-provide-composable';
export * from './input-composable';
export * from './input-form-composable';
export * from './input-modifier-composable';
export * from './input-option-composable';
export * from './responsive-composable';
export * from './style-set-composable';
export * from './state-class-composable';
export * from './value-matcher-composable';
