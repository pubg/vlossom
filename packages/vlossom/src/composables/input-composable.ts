import { ComputedRef, PropType, Ref, computed, nextTick, onMounted, ref, watch } from 'vue';
import { StateMessage, Rule, Message, UIState, InputComponentOptions } from '@/declaration/types';

export function getInputProps<T = unknown>() {
    return {
        disabled: { type: Boolean, default: false },
        label: { type: String, default: '' },
        messages: { type: Array as PropType<Message<T>[]>, default: () => [] },
        noClear: { type: Boolean, default: false },
        noLabel: { type: Boolean, default: false },
        noMsg: { type: Boolean, default: false },
        placeholder: { type: String, default: '' },
        readonly: { type: Boolean, default: false },
        required: { type: Boolean, default: false },
        rules: { type: Array as PropType<Rule<T>[]>, default: () => [] },
        visible: { type: Boolean, default: true },

        // v-model
        changed: { type: Boolean, default: false },
        valid: { type: Boolean, default: false },
    };
}

export function useInput<T = unknown>(
    inputValue: Ref<T>,
    modelValue: Ref<T>,
    ctx: any,
    options?: InputComponentOptions<T>,
) {
    const { emit } = ctx;
    const { messages = ref([]), rules = ref([]) } = options || {};

    const changed = ref(false);
    const isInitialized = ref(false);

    const innerMessages: Ref<StateMessage[]> = ref([]);
    async function checkMessages() {
        innerMessages.value = [];
        const pendingMessages: Promise<StateMessage>[] = [];
        messages.value.forEach((message) => {
            if (typeof message === 'function') {
                const result = message(inputValue.value);
                if (result instanceof Promise) {
                    pendingMessages.push(result);
                } else {
                    innerMessages.value.push(result as StateMessage);
                }
            } else {
                innerMessages.value.push(message);
            }
        });

        if (pendingMessages.length === 0) {
            return;
        }
        const resolvedMessages = await Promise.all(pendingMessages);
        innerMessages.value.push(...resolvedMessages);
    }
    watch(messages, checkMessages, { deep: true });

    const showRuleMessages = ref(false);
    const ruleMessages: Ref<StateMessage[]> = ref([]);
    async function checkRules() {
        ruleMessages.value = [];

        const pendingRules: Promise<string>[] = [];
        rules.value.forEach((rule) => {
            const result = rule(inputValue.value);
            if (!result) {
                return;
            }
            if (result instanceof Promise) {
                pendingRules.push(result);
            } else {
                ruleMessages.value.push({ state: UIState.DANGER, message: result as string });
            }
        });

        if (pendingRules.length === 0) {
            return;
        }
        const resolvedMessages = (await Promise.all(pendingRules)).map((resolved) => ({
            state: UIState.DANGER,
            message: resolved,
        }));
        ruleMessages.value.push(...resolvedMessages);
    }
    watch(rules, checkRules, { deep: true });

    const computedMessages: ComputedRef<StateMessage[]> = computed(() => {
        if (showRuleMessages.value) {
            return [...innerMessages.value, ...ruleMessages.value];
        }

        return innerMessages.value;
    });

    watch(inputValue, (value, oldValue) => {
        emit('update:modelValue', value);
        if (options?.callbacks?.onChange) {
            options.callbacks.onChange(value, oldValue);
        }

        checkMessages();
        checkRules();

        if (!isInitialized.value) {
            return;
        }
        changed.value = true;
        showRuleMessages.value = true;
        emit('change', value);
    });

    watch(
        modelValue,
        (value) => {
            inputValue.value = value;
        },
        { deep: true },
    );

    watch(changed, () => {
        emit('update:changed', changed.value);
    });

    const valid = computed(() => ruleMessages.value.length === 0);
    watch(valid, () => {
        emit('update:valid', valid.value);
    });

    onMounted(() => {
        inputValue.value = modelValue.value;
        checkMessages();
        checkRules();

        if (options?.callbacks?.onMounted) {
            options.callbacks?.onMounted();
        }
        nextTick(() => {
            isInitialized.value = true;
        });
    });

    function validate(): boolean {
        showRuleMessages.value = true;
        return ruleMessages.value.length === 0;
    }

    return {
        changed,
        valid,
        computedMessages,
        showRuleMessages,
        validate,
    };
}
