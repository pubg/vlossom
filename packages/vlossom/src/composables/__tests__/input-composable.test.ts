import { beforeEach, afterEach, describe, expect, it, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import { defineComponent, nextTick, ref, toRefs } from 'vue';
import { getInputProps, useInput } from '@/composables/input-composable';
import { StateMessage, UIState } from '@/declaration/types';

describe('input composable', () => {
    let onMountedSpy = vi.fn();
    let onChangeSpy = vi.fn();
    const inputValue = ref('');

    const InputComponent = defineComponent({
        render: () => null,
        props: {
            modelValue: { type: String, default: '' },
            ...getInputProps<string>(),
        },
        setup(props, ctx) {
            const { modelValue, messages, rules } = toRefs(props);

            return {
                ...useInput(inputValue, modelValue, ctx, {
                    callbacks: {
                        onMounted: onMountedSpy,
                        onChange: onChangeSpy,
                    },
                    messages,
                    rules,
                }),
            };
        },
    });

    beforeEach(() => {
        inputValue.value = '';
        onMountedSpy = vi.fn();
        onChangeSpy = vi.fn();
    });

    afterEach(() => {
        vi.restoreAllMocks();
    });

    describe('value binding (inputValue, modelValue)', () => {
        it('modelValue 초깃값이 있으면 inputValue가 업데이트 된다', async () => {
            // when
            const wrapper = mount(InputComponent, {
                props: {
                    modelValue: 'test',
                    'onUpdate:modelValue': (v: string) => wrapper.setProps({ modelValue: v }),
                },
            });

            // then
            expect(inputValue.value).toBe('test');
            expect(wrapper.vm.changed).toBe(false);
            expect(wrapper.vm.modelValue).toBe('test');
        });

        it('inputValue 값을 업데이트 해서 modelValue를 업데이트 할 수 있다', async () => {
            // given
            const wrapper = mount(InputComponent, {
                props: {
                    modelValue: '',
                    'onUpdate:modelValue': (v: string) => wrapper.setProps({ modelValue: v }),
                },
            });

            // when
            await nextTick();
            inputValue.value = 'test';
            await nextTick();

            // then
            expect(inputValue.value).toBe('test');
            expect(wrapper.vm.modelValue).toBe('test');
            expect(wrapper.vm.changed).toBe(true);
            expect(wrapper.emitted('update:modelValue')?.[0]).toEqual(['test']);
        });

        it('modelValue를 바꿔서 inputValue 값을 바꿀 수 있다', async () => {
            // given
            const wrapper = mount(InputComponent, {
                props: {
                    modelValue: '',
                    'onUpdate:modelValue': (v: string) => wrapper.setProps({ modelValue: v }),
                },
            });

            // when
            await nextTick();
            await wrapper.setProps({ modelValue: 'test' });

            // then
            expect(inputValue.value).toBe('test');
            expect(wrapper.vm.modelValue).toBe('test');
            expect(wrapper.vm.changed).toBe(true);
            expect(wrapper.emitted().change).toHaveLength(1);
            expect(wrapper.emitted().change?.[0]).toEqual(['test']);
        });

        it('mount 시점에 onMounted callback을 실행한다', () => {
            // given
            // when
            const wrapper = mount(InputComponent, {
                props: {
                    modelValue: '',
                    'onUpdate:modelValue': (v: string) => wrapper.setProps({ modelValue: v }),
                },
            });

            // then
            expect(onMountedSpy).toHaveBeenCalledTimes(1);
        });

        it('inputValue가 변경되면 onChange callback을 실행한다', async () => {
            // given
            const wrapper = mount(InputComponent, {
                props: {
                    modelValue: '',
                    'onUpdate:modelValue': (v: string) => wrapper.setProps({ modelValue: v }),
                },
            });

            // when
            await nextTick();
            inputValue.value = 'test';
            await nextTick();

            // then
            expect(onChangeSpy).toHaveBeenCalledTimes(1);
            expect(onChangeSpy).toHaveBeenCalledWith('test', '');
        });
    });

    describe('messages', () => {
        it('messages를 StateMessage[] 형태로 전달할 수 있다', async () => {
            // given
            const wrapper = mount(InputComponent, {
                props: {
                    messages: [
                        { state: UIState.INFO, message: 'info message' },
                        { state: UIState.SUCCESS, message: 'success message' },
                        { state: UIState.WARN, message: 'warning message' },
                    ],
                },
            });

            // then
            expect(wrapper.vm.computedMessages).toHaveLength(3);
            expect(wrapper.vm.computedMessages[0].state).toBe(UIState.INFO);
            expect(wrapper.vm.computedMessages[0].message).toBe('info message');
        });

        it('messages를 함수로 전달할 수 있다', () => {
            // given
            const wrapper = mount(InputComponent, {
                props: {
                    messages: [
                        () => ({ state: UIState.INFO, message: 'info message' }),
                        () => ({ state: UIState.SUCCESS, message: 'success message' }),
                        () => ({ state: UIState.WARN, message: 'warning message' }),
                    ],
                },
            });

            // then
            expect(wrapper.vm.computedMessages).toHaveLength(3);
            expect(wrapper.vm.computedMessages[0].state).toBe(UIState.INFO);
            expect(wrapper.vm.computedMessages[0].message).toBe('info message');
        });

        it('messages를 PromiseLike를 반환하는 함수로도 전달할 수 있다', async () => {
            // given
            const wrapper = mount(InputComponent, {
                props: {
                    messages: [
                        () => Promise.resolve({ state: UIState.INFO, message: 'info message' }),
                        () => Promise.resolve({ state: UIState.SUCCESS, message: 'success message' }),
                        () => Promise.resolve({ state: UIState.WARN, message: 'warning message' }),
                    ],
                },
            });

            // call nextTick twice to check PromiseLike rule
            await nextTick();
            await nextTick();

            // then
            expect(wrapper.vm.computedMessages).toHaveLength(3);
            expect(wrapper.vm.computedMessages[0].state).toBe(UIState.INFO);
            expect(wrapper.vm.computedMessages[0].message).toBe('info message');
        });

        it('messages가 바뀌면 바뀐 message를 반영할 수 있다', async () => {
            // given
            const wrapper = mount(InputComponent, {
                props: {
                    messages: [
                        { state: UIState.INFO, message: 'info message' },
                        { state: UIState.SUCCESS, message: 'success message' },
                        { state: UIState.WARN, message: 'warning message' },
                    ],
                },
            });

            await wrapper.setProps({
                messages: [{ state: UIState.DANGER, message: 'changed message' }],
            });

            // then
            expect(wrapper.vm.computedMessages).toHaveLength(1);
            expect(wrapper.vm.computedMessages[0].state).toBe(UIState.DANGER);
            expect(wrapper.vm.computedMessages[0].message).toBe('changed message');
        });
    });

    describe('rules', () => {
        it('rule이 설정되었어도 값의 변경이 없다면 message가 없다', () => {
            //given
            const wrapper = mount(InputComponent, {
                props: {
                    rules: [(v: string) => (v ? '' : 'required')],
                },
            });

            // then
            expect(wrapper.vm.valid).toBe(false);
            expect(wrapper.vm.changed).toBe(false);
            expect(wrapper.vm.showRuleMessages).toBe(false);
            expect(wrapper.vm.computedMessages).toHaveLength(0);
        });

        it('값이 유효하면 rule error message가 없다', async () => {
            //given
            const wrapper = mount(InputComponent, {
                props: {
                    rules: [(v: string) => (v ? '' : 'required')],
                },
            });

            // when
            await nextTick();
            inputValue.value = 'test';
            await nextTick();

            // then
            expect(wrapper.vm.valid).toBe(true);
            expect(wrapper.vm.changed).toBe(true);
            expect(wrapper.vm.showRuleMessages).toBe(true);
            expect(wrapper.vm.computedMessages).toHaveLength(0);
        });

        it('값이 유효하지 않으면 rule error message가 있다', async () => {
            //given
            const wrapper = mount(InputComponent, {
                props: {
                    rules: [(v: string) => (v ? '' : 'required')],
                },
            });

            // when
            await nextTick();
            inputValue.value = 'test';
            await nextTick();
            inputValue.value = '';
            await nextTick();

            // then
            expect(wrapper.vm.valid).toBe(false);
            expect(wrapper.vm.changed).toBe(true);
            expect(wrapper.vm.showRuleMessages).toBe(true);
            expect(wrapper.vm.computedMessages).toEqual([{ state: UIState.DANGER, message: 'required' }]);
        });

        it('PromiseLike의 rule도 체크할 수 있다', async () => {
            //given
            const wrapper = mount(InputComponent, {
                props: {
                    rules: [(v: string) => Promise.resolve(v ? '' : 'required')],
                },
            });

            // when
            await nextTick();
            inputValue.value = 'test';
            await nextTick();
            inputValue.value = '';
            // call nextTick twice to check PromiseLike rule
            await nextTick();
            await nextTick();

            // then
            expect(wrapper.vm.valid).toBe(false);
            expect(wrapper.vm.changed).toBe(true);
            expect(wrapper.vm.showRuleMessages).toBe(true);
            expect(wrapper.vm.computedMessages).toEqual([{ state: UIState.DANGER, message: 'required' }]);
        });

        it('기존 message가 있으면 rule 체크 결과를 danger 타입으로 추가한다', async () => {
            // given
            const infoMsg: StateMessage = { state: UIState.INFO, message: 'info message' };
            const wrapper = mount(InputComponent, {
                props: {
                    messages: [infoMsg],
                    rules: [(v: string) => (v ? '' : 'required')],
                },
            });

            // when
            await nextTick();
            inputValue.value = 'test';
            await nextTick();
            inputValue.value = '';
            await nextTick();

            // then
            expect(wrapper.vm.valid).toBe(false);
            expect(wrapper.vm.changed).toBe(true);
            expect(wrapper.vm.showRuleMessages).toBe(true);
            expect(wrapper.vm.computedMessages).toHaveLength(2);
            expect(wrapper.vm.computedMessages[0]).toEqual(infoMsg);
            expect(wrapper.vm.computedMessages[1]).toEqual({ state: UIState.DANGER, message: 'required' });
        });
    });

    describe('validate', () => {
        it('validate 함수를 호출하면 변경이 없어도 message가 노출된다', async () => {
            //given
            const wrapper = mount(InputComponent, {
                props: {
                    rules: [(v: string) => (v ? '' : 'required')],
                },
            });

            // when
            const result = wrapper.vm.validate();
            await nextTick();

            // then
            expect(result).toBe(false);
            expect(wrapper.vm.valid).toBe(false);
            expect(wrapper.vm.changed).toBe(false);
            expect(wrapper.vm.showRuleMessages).toBe(true);
            expect(wrapper.vm.computedMessages).toEqual([{ state: UIState.DANGER, message: 'required' }]);
        });

        describe('shake', () => {
            it('validate를 호출 했을 때 값이 유효하면 shake 값(boolean)에 변화가 없다', async () => {
                // given
                const wrapper = mount(InputComponent, {
                    props: {
                        rules: [(v: string) => (v ? '' : 'required')],
                    },
                });
                const oldShake = wrapper.vm.shake;
                await nextTick();
                inputValue.value = 'test';
                await nextTick();

                // when
                const result = wrapper.vm.validate();
                await nextTick();

                // then
                expect(result).toBe(true);
                expect(wrapper.vm.shake).toBe(oldShake);
            });

            it('validate를 호출 했을 때 값이 유효하지 않으면 shake 값(boolean)이 바뀐다', async () => {
                // given
                const wrapper = mount(InputComponent, {
                    props: {
                        rules: [(v: string) => (v ? '' : 'required')],
                    },
                });
                const oldShake = wrapper.vm.shake;

                // when
                const result = wrapper.vm.validate();
                await nextTick();

                // then
                expect(result).toBe(false);
                expect(wrapper.vm.shake).toBe(!oldShake);
            });
        });
    });
});
