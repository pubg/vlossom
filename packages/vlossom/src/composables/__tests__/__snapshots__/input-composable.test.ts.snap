// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getInputProps > input component에 필요한 props들을 가져올 수 있다 1`] = `
{
  "ariaLabel": {
    "default": null,
    "type": [Function],
  },
  "changed": {
    "default": false,
    "type": [Function],
  },
  "disabled": {
    "default": false,
    "type": [Function],
  },
  "id": {
    "default": "",
    "type": [Function],
  },
  "label": {
    "default": "",
    "type": [Function],
  },
  "messages": {
    "default": [Function],
    "type": [Function],
  },
  "name": {
    "default": "",
    "type": [Function],
  },
  "noClear": {
    "default": false,
    "type": [Function],
  },
  "noDefaultRules": {
    "default": false,
    "type": [Function],
  },
  "noMessage": {
    "default": false,
    "type": [Function],
  },
  "placeholder": {
    "default": "",
    "type": [Function],
  },
  "readonly": {
    "default": false,
    "type": [Function],
  },
  "required": {
    "default": false,
    "type": [Function],
  },
  "rules": {
    "default": [Function],
    "type": [Function],
  },
  "state": {
    "default": "idle",
    "type": [Function],
  },
  "valid": {
    "default": false,
    "type": [Function],
  },
  "visible": {
    "default": true,
    "type": [Function],
  },
}
`;

exports[`getInputProps > input props 중 제외할 props를 정할 수 있다 1`] = `
{
  "ariaLabel": {
    "default": null,
    "type": [Function],
  },
  "changed": {
    "default": false,
    "type": [Function],
  },
  "disabled": {
    "default": false,
    "type": [Function],
  },
  "id": {
    "default": "",
    "type": [Function],
  },
  "name": {
    "default": "",
    "type": [Function],
  },
  "noClear": {
    "default": false,
    "type": [Function],
  },
  "noDefaultRules": {
    "default": false,
    "type": [Function],
  },
  "noMessage": {
    "default": false,
    "type": [Function],
  },
  "placeholder": {
    "default": "",
    "type": [Function],
  },
  "readonly": {
    "default": false,
    "type": [Function],
  },
  "required": {
    "default": false,
    "type": [Function],
  },
  "rules": {
    "default": [Function],
    "type": [Function],
  },
  "state": {
    "default": "idle",
    "type": [Function],
  },
  "valid": {
    "default": false,
    "type": [Function],
  },
  "visible": {
    "default": true,
    "type": [Function],
  },
}
`;
